entity-service-impl/src/main/java/org/hypertrace/entity/query/service/EntityQueryServiceImpl.java

```diff
@@ -290,19 +291,128 @@ private void doUpdate(RequestContext requestContext, EntityUpdateRequest request
             new SingleValueKey(requestContext.getTenantId().orElseThrow(), entityId);
         // TODO better error reporting once doc store exposes the,
-        if (!entitiesCollection.updateSubDoc(key, subDocPath, new JSONDocument(jsonValue))) {
-          LOG.warn(
-              "Failed to update entity {}, subDocPath {}, with new doc {}.",
-              key,
-              subDocPath,
-              jsonValue);
+        if (entitiesUpdateMap.containsKey(key)) {
+          entitiesUpdateMap.get(key).put(subDocPath, new JSONDocument(jsonValue));
+        } else {
+          Map<String, Document> subDocument = new HashMap<>();
+          subDocument.put(subDocPath, new JSONDocument(jsonValue));
+          entitiesUpdateMap.put(key, subDocument);
         }
       }
+      try {
+        entitiesCollection.bulkUpdateSubDocs(entitiesUpdateMap);
+      } catch (Exception e) {
+        LOG.warn(
+            "Failed to update entities, subDocPath {}, with new doc {}.", subDocPath, jsonValue);
+      }
+    }
+  }
+
+  @Override
+  public void bulkUpdate(
+      BulkEntityUpdateRequest request, StreamObserver<ResultSetChunk> responseObserver) {
+    // Validations
+    RequestContext requestContext = RequestContext.CURRENT.get();
+    Optional<String> tenantId = requestContext.getTenantId();
+    if (tenantId.isEmpty()) {
+      responseObserver.onError(new ServiceException("Tenant id is missing in the request."));
+      return;
+    }
+    if (StringUtils.isEmpty(request.getEntityType())) {
+      responseObserver.onError(new ServiceException("Entity type is missing in the request."));
+      return;
+    }
+    if (request.getEntitiesCount() == 0) {
+      responseObserver.onError(new ServiceException("Entity IDs are missing in the request."));
+    }
+
+    try {
+      doBulkUpdate(requestContext, request);
+      // Finally return the selections
+      Set<String> entityIdsList = request.getEntitiesMap().keySet();
+      returnSelections(
+          entityIdsList, request.getSelectionList(), requestContext, tenantId, responseObserver);
+    } catch (Exception e) {
+      responseObserver.onError(
+          new ServiceException("Error occurred while executing " + request, e));
+    }
+  }
+
+  private void returnSelections(
+      Iterable<String> entityIdsList,
+      List<Expression> selectionList,
+      RequestContext requestContext,
+      Optional<String> tenantId,
+      StreamObserver<ResultSetChunk> responseObserver)
+      throws Exception {
+    Query entitiesQuery = Query.newBuilder().addAllEntityId(entityIdsList).build();
+    List<String> docStoreSelections =
+        entityQueryConverter.convertSelectionsToDocStoreSelections(requestContext, selectionList);
+    Iterator<Document> documentIterator =
+        entitiesCollection.search(
+            DocStoreConverter.transform(tenantId.get(), entitiesQuery, docStoreSelections));
+    List<Entity> entities = convertDocsToEntities(documentIterator);
+    responseObserver.onNext(
```

Comment:
This shouldn't have the response observer at all. This should only return the list of entities. The consumers should take care of converting the entities into result set chunks and sending them via response observer